lib/auth.js

```diff
@@ -0,0 +1,470 @@
+    name: '_json.info.user.real_name',
+    provider: 'slack'
+  }, site)));
+}
+
+/**
+ * verify LDAP auth a little differently
+ * note: this basically wraps verify with a different function signature
+ * @param {object} site
+ * @returns {function}
+ */
+function verifyLdap(site) {
+  return function (req, user, done) {
+    // the callback for LDAP is different than oauth, so we need to
+    // pass different options to verify()
+    module.exports.verify({ // allows this to be mocked in tests
+      username: 'sAMAccountName',
+      imageUrl: '', // ldap generally has no images
+      name: 'displayName',
+      provider: 'ldap'
+    }, site)(req, null, null, user, done); // eslint-disable-line
+  };
+}
+
+/**
+ * ldap / active directory auth strategy
+ * @param {object} site
+ */
+function createLDAPStrategy(site) {
+  const LDAPStrategy = require('passport-ldapauth');
+
+  passport.use(`ldap-${site.slug}`, new LDAPStrategy({
+    server: {
+      url: process.env.LDAP_URL,
+      adminDn: process.env.LDAP_BIND_DN,
+      adminPassword: process.env.LDAP_BIND_CREDENTIALS,
+      searchBase: process.env.LDAP_SEARCH_BASE,
+      searchFilter: process.env.LDAP_SEARCH_FILTER
+    },
+    passReqToCallback: true,
+    credentialsLookup: basicAuth,
+  }, verifyLdap(site)));
+}
+
+/**
+ * api key callback, checks to see if api key provided matches env variable
+ * @param {string} apikey
+ * @param {function} done
+ */
+function apiCallback(apikey, done) {
+  if (apikey === process.env.CLAY_ACCESS_KEY) {
+    done(null, { provider: 'apikey' });
+  } else {
+    done(null, false, { message: 'Unknown apikey: ' + apikey });
+  }
+}
+
+/**
+ * api key strategy
+ * matches against the CLAY_ACCESS_KEY env variable
+ * @param {object} site
+ */
+function createAPIKeyStrategy() {
+  const APIKeyStrategy = require('passport-http-header-token').Strategy;
+
+  passport.use('apikey', new APIKeyStrategy({}, apiCallback));
+}
+
+/**
+ * create the specified provider strategy
+ * @param {object} site
+ * @throws {Error} if unsupported strategy
+ * @returns {Function}
+ */
+function createStrategy(site) {
+  return function (provider) {
+    switch (provider) {
+      case 'twitter': createTwitterStrategy(site);
+        // twitter requires TWITTER_CONSUMER_KEY and TWITTER_CONSUMER_SECRET env variables
+        break;
+      case 'google': createGoogleStrategy(site);
+        // google requires GOOGLE_CONSUMER_KEY and GOOGLE_CONSUMER_SECRET
+        break;
+      case 'slack': createSlackStrategy(site);
+        // slack requires SLACK_CONSUMER_KEY and SLACK_CONSUMER_SECRET
+        break;
+      case 'ldap': createLDAPStrategy(site);
+        // ldap requires LDAP_URL, LDAP_BIND_DN, LDAP_BIND_CREDENTIALS, LDAP_SEARCH_BASE, and LDAP_SEARCH_FILTER
+        break;
+      case 'apikey': createAPIKeyStrategy();
+        // apikey requires CLAY_ACCESS_KEY
+        break;
+      default: throw new Error(`Unknown provider: ${provider}!`);
+    }
+  };
+}
```

Comment:
Is it worth it creating a providers folder and just dynamically import provider's strategy creator as well as any other provider-specific settings from there based on provider name?  Simplifies this and moves all those domain-specific logic to their own containers.